### ISSUE TITLE
Deduplication Error

### ISSUE THREAD

**Issue by gxander85** [boredazfcuk/docker-icloudpd#384]
Over the last few weeks I have been using this service. I recently posted an issue that I closed, but I have a definitive issue that needs investigation in this space.

I have been seeing deduplication logs for images that are 100% not duplicates.

I have set folder_structure to {:%Y/%Y-%m/%Y-%m-%d}, so I should only ever get a deduplication when the image is saved into the folder for the same day. Correct?

Example:

Library contains IMG_1977.DNG 3 times and one IMG_1977.JPG.
Each of these are on different dates 2022.02.20, 2022.11.21 & 2023.02.23 for .dng and 2023.07.31 for .jpg
Within the download directories, only one file has been downloaded
/iCloud/2023/2023-07/2023-07-31/IMG_1977-287462.JPG

Note the additional suffix on the image of -287462

In the directory, there are two versions of this file.
IMG_1977.JPG
IMG_1977-287462.JPG

Below are the logs that relate to IMG_1977

{log:/iCloud/2023/2023-07/2023-07-31/IMG_1977.JPG\r\n,stream:stdout,time:2023-08-29T02:07:52.279760542Z}
{log:/iCloud/2022/2022-11/2022-11-21/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T02:33:08.333522065Z}
{log:/iCloud/2022/2022-02/2022-02-20/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T02:40:01.223690965Z}
{log:/iCloud/2023/2023-02/2023-02-23/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T03:10:02.834070843Z}
{log:2023-08-29 19:31:37 INFO     Downloading /iCloud/2023/2023-07/2023-07-31/IMG_1977.JPG\r\n,stream:stdout,time:2023-08-29T09:31:37.574252858Z}
{log:2023-08-29 19:31:38 DEBUG    Setting EXIF timestamp for /iCloud/2023/2023-07/2023-07-31/IMG_1977.JPG: 2023-07-31 17:43:22+1000\r\n,stream:stdout,time:2023-08-29T09:31:38.221242063Z}
{log:/iCloud/2023/2023-07/2023-07-31/IMG_1977-287462.JPG\r\n,stream:stdout,time:2023-08-29T10:19:34.295454525Z}
{log:/iCloud/2022/2022-11/2022-11-21/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T10:44:55.385595122Z}
{log:/iCloud/2022/2022-02/2022-02-20/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T10:51:29.647382421Z}
{log:/iCloud/2023/2023-02/2023-02-23/IMG_1977.DNG\r\n,stream:stdout,time:2023-08-29T11:21:30.900476546Z}
{log:2023-08-29 21:34:26 INFO     /iCloud/2023/2023-07/2023-07-31/IMG_1977-287462.JPG deduplicated.\r\n,stream:stdout,time:2023-08-29T11:34:26.310348993Z}
{log:2023-08-29 21:34:26 INFO     Downloading /iCloud/2023/2023-07/2023-07-31/IMG_1977-287462.JPG\r\n,stream:stdout,time:2023-08-29T11:34:26.311067778Z}
{log:2023-08-29 21:34:26 DEBUG    Setting EXIF timestamp for /iCloud/2023/2023-07/2023-07-31/IMG_1977-287462.JPG: 2023-07-31 17:43:22+1000\r\n,stream:stdout,time:2023-08-29T11:34:26.86641874Z}

There is no record of anything but the JPG being downloaded, and you can see it has actually been downloaded twice.

**Comment by gxander85** [comment]
Further to this, I have seen instances where there is 100% only one file in iCloud photos, but the service has identified it as deduplicated. These are typically images/videos saved from Instagram when making posts, stories or reels.

Example, was parsed once on the first sync, then appears to be parsed again on the next sync.

{log:/iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32.jpg\r\n,stream:stdout,time:2023-08-29T02:00:07.071986504Z}
{log:2023-08-29 15:51:15 INFO     Downloading /iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32.jpg\r\n,stream:stdout,time:2023-08-29T05:51:15.14611468Z}
{log:2023-08-29 15:51:15 DEBUG    Setting EXIF timestamp for /iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32.jpg: 2023-08-20 18:55:56+1000\r\n,stream:stdout,time:2023-08-29T05:51:15.694008405Z}
{log:/iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32-156187.jpg\r\n,stream:stdout,time:2023-08-29T10:19:34.295140993Z}
{log:2023-08-29 21:25:52 INFO     /iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32-156187.jpg deduplicated.\r\n,stream:stdout,time:2023-08-29T11:25:52.739818606Z}
{log:2023-08-29 21:25:52 INFO     Downloading /iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32-156187.jpg\r\n,stream:stdout,time:2023-08-29T11:25:52.74056307Z}
{log:2023-08-29 21:25:53 DEBUG    Setting EXIF timestamp for /iCloud/2023/2023-08/2023-08-20/1BA0C0BA-4D5F-40A1-95FB-5B1131B2CC32-156187.jpg: 2023-08-20 18:55:56+1000\r\n,stream:stdout,time:2023-08-29T11:25:53.263528582Z}

**Comment by boredazfcuk** [comment]
This is a side effect of this:

DEBUG Setting EXIF timestamp for /iCloud/2023/2023-07/2023-07-31/IMG_1977.JPG

When the EXIF data is used to set the time stamp, the file is then different from the original, but has the same name. 

On the next download run, it will detect that the original needs downloading again, as there is no match on disk, downloads the file, but has to append a dedupe tag due to the modified version already being there.

I've switched this off on my own instance, as for the reason you point out, it only occurs on files downloaded from the internet, or saved from applications, etc.

**Comment by boredazfcuk** [comment]
Also, I have actually reported this as a bug in the upstream application a couple of years ago:

https://github.com/icloud-photos-downloader/icloud_photos_downloader/issues/380

**Comment by gxander85** [comment]
> This is a side effect of this:
> 
> DEBUG Setting EXIF timestamp for /iCloud/2023/2023-07/2023-07-31/IMG_1977.JPG

Does this mean that I will forever be in a de-dupe loop with this option set?

If the file save from Instagram is in the library as 2023-08-30 and the process downloads the file tomorrow on next sync, will the date stamp be as it is in library? Or be the date it is download as no EXIF has been set?

**Comment by boredazfcuk** [comment]
So, say you have IMG_0000.JPG with date 1/1/2023 as the file's time stamp, and empty EXIF date, in your iCloud Photos.

The first run will download this file. It will see there is empty EXIF data in the file and set the EXIF date to be 1/1/2023.

The second run will see the file IMG_000.JPG, with date 1/1/2023 and empty EXIF date in the cloud and search for a match on the local disk. It will not find a match, so it will download the file. However, because a file named IMG_000.JPG already exists in the download folder, it has to save the file with a different name. It will then look at the size of the file, say 1,234,567,890 bytes and use the last 6 digits of that as the dedupe name, IMG_000-567890.JPG.

From the third run onwards, it will see the file IMG_000.JPG, with date 1/1/2023 and empty EXIF date in the cloud and search for a match on the local disk. It will then see the file is a match for IMG_0000-567890.JPG and skip downloading the file.

**Comment by gxander85** [comment]
So what you're saying is it will download the original & set the EXIF. Then download it again, with the dedupe file name. Then it won't download it again? I'll just have two local files saved. Both essential the same, just with an append file name?

And all of this runaround is a byproduct of updating the EXIF?

Here I was thinking my library had duplicate files, and I was searching and only finding one. Thought I was going mad.

**Comment by boredazfcuk** [comment]
> So what you're saying is it will download the original & set the EXIF. Then download it again, with the dedupe file name. Then it won't download it again? I'll just have two local files saved. Both essential the same, just with an append file name?

Yes, exactly this.

> And all of this runaround is a byproduct of updating the EXIF?

Yes

> Here I was thinking my library had duplicate files, and I was searching and only finding one. Thought I was going mad.

Same. Once I'd figured out it was the EXIF setting, I just recursively deleted all .JPG files, switched off the variable, and re-ran the script.

**Comment by gxander85** [comment]
I'll give this all a shot after the download process has completed.

I have 80k+ files and it's been a long road to almost get everything the way I want it.

I'll report back in a few days

**Comment by gxander85** [comment]
Okay, so, I have reset my configuration and set the 'set_exif' option to false. Cleaned out the directory of all the deduped files just by removing the day folders. Rerun the sync and all seems to be working as you mention. It is actually exactly what I wanted to achieve.

However, I have noticed that video files that I have trimmed and 'saved as new clip' end up being deduplicated. This may be the intention of the service, but wondering if there was a different way the dedupe config can handle these? All clips have the same filename, but are different duration. Maybe the current process is the only way to handle these, I am not sure.

Also, curious that the extension case sensitivity didn't trigger an additional dedupe.

{log:/iCloud/2023/2023-05/2023-05-07/IMG_6860.mov\r\n,stream:stdout,time:2023-08-29T11:18:12.972681105Z}
{log:/iCloud/2023/2023-05/2023-05-07/IMG_6860.mov\r\n,stream:stdout,time:2023-08-29T11:18:21.552252771Z}
{log:/iCloud/2023/2023-05/2023-05-07/IMG_6860.MOV\r\n,stream:stdout,time:2023-08-29T11:18:21.552365797Z}
{log:2023-09-02 02:55:32 INFO     Downloading /iCloud/2023/2023-05/2023-05-07/IMG_6860.mov\r\n,stream:stdout,time:2023-09-01T16:55:32.230429811Z}
{log:2023-09-02 03:07:50 INFO     /iCloud/2023/2023-05/2023-05-07/IMG_6860-180911341.mov deduplicated.\r\n,stream:stdout,time:2023-09-01T17:07:50.852403442Z}
{log:2023-09-02 03:07:50 INFO     Downloading /iCloud/2023/2023-05/2023-05-07/IMG_6860-180911341.mov\r\n,stream:stdout,time:2023-09-01T17:07:50.852542386Z}
{log:2023-09-02 03:08:11 INFO     Downloading /iCloud/2023/2023-05/2023-05-07/IMG_6860.MOV\r\n,stream:stdout,time:2023-09-01T17:08:11.869614812Z}

I have lots of these as I film concert content and then post to YouTube. I have got around most of these by captioning the video file with the title of the clip, exporting, deleting, then importing so the filename is different. I will likely continue this as it is great for archival purposes in the long term.

In short, if the dedupe process is the only way to handle these trimmed/new clip files, then I will accept it and we can close this out.

Thank you again for the insight into how this all works. Its a fantastic service.

**Comment by boredazfcuk** [comment]
I'm not sure how else it can be handled to be honest... The filenames must be unique, so they have to be changed somehow. The script also needs to b able to workout which deduped file is belongs to which photo in icloud. It also cannot make any changes to the file or it will trigger de-dupe again.

I think the current method has been show to work quite well, even though it doesn't guarantee the file name is unique, it's still unique enough that I haven't seen anyone report an issue. I believe it has been mentioned that it could be changed to append the file hash instead of file size, but that would make the filenames quite long and introduce quite a bit of overhead, as hashing a video file isn't instant. The other option would be for icloudpd to keep a database, but that would mean a significant re-write of the application and I'm not sure the developers have the time or appetite for such a large change.
